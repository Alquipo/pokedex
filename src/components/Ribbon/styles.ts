import { darken } from 'polished'
import styled, { css, DefaultTheme } from 'styled-components'
import { RibbonColors, RibbonProps } from '.'

const wrapperModifiers = {
  color: (theme: DefaultTheme, color: RibbonColors) => css`
    background-color: ${theme.colors.pokemonType[color]};

    &::before {
      border-left-color: ${darken(0.2, theme.colors.pokemonType[color])};
      border-top-color: ${darken(0.2, theme.colors.pokemonType[color])};
    }
  `,

  right: () => css`
    right: -1.5rem;

    &::before {
      border-right-width: 1.5rem;
      border-left-width: 0rem;
      border-right-color: transparent;
      right: 0;
    }
  `,

  left: () => css`
    left: -1.5rem;

    &::before {
      border-left-width: 1.5rem;
      border-right-width: 0rem;
      border-left-color: transparent;
      left: 0;
    }
  `
}

export const Wrapper = styled.div<RibbonProps>`
  ${({ theme, color, position }) => css`
    position: absolute;

    padding: 0 ${theme.spacings.xsmall};
    height: 2.6rem;

    top: ${theme.spacings.xxsmall};
    display: flex;
    align-items: center;

    font-size: ${theme.font.sizes.xsmall};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.white};

    &::before {
      content: '';
      top: 2.6rem;
      border-top-width: 0.7rem;
      position: absolute;
      border-style: solid;
      border-bottom-color: transparent;
      border-bottom-width: 1rem;
    }

    ${!!color && wrapperModifiers.color(theme, color)};
    ${!!position && wrapperModifiers[position]()};
  `}
`
