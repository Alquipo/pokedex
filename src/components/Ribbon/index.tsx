import { PokemonTypesProps } from 'components/PokemonCard'
import * as S from './styles'

export type RibbonColors = PokemonTypesProps
export type RibbonPosition = 'right' | 'left'

export type RibbonProps = {
  children: React.ReactNode
  color?: RibbonColors
  position?: RibbonPosition
}
const Ribbon = ({ children, color, position = 'left' }: RibbonProps) => (
  <S.Wrapper position={position} color={color}>
    {children}
  </S.Wrapper>
)

export default Ribbon
