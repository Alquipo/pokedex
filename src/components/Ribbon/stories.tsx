import { Story, Meta } from '@storybook/react/types-6-0'
import Ribbon, { RibbonProps } from '.'

export default {
  title: 'Home/Ribbon',
  component: Ribbon,
  args: {
    children: '#005',
    color: 'normal'
  },
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<RibbonProps> = (args) => (
  <div
    style={{
      height: '16rem',
      width: '18rem ',

      position: 'relative',

      padding: '0.3rem',
      margin: '0.3rem 0.8rem',

      borderRadius: '1rem',
      border: '0.1rem solid',
      textAlign: 'center',

      cursor: 'default',

      backgroundColor: 'white'
    }}
  >
    <Ribbon {...args} />
  </div>
)
