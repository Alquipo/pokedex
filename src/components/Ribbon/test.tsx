import { render, screen } from '@testing-library/react'

import Ribbon from '.'

describe('<Ribbon />', () => {
  it('should render the heading', () => {
    const { container } = render(<Ribbon />)

    expect(screen.getByRole('heading', { name: /Ribbon/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
