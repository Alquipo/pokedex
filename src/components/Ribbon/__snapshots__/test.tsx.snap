// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ribbon /> should render the text correctly 1`] = `
.c0 {
  position: absolute;
  padding: 0 0.8rem;
  height: 2.6rem;
  top: 0.3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.2rem;
  font-weight: 700;
  color: #fff;
  left: -1.5rem;
}

.c0::before {
  content: '';
  top: 2.6rem;
  border-top-width: 0.7rem;
  position: absolute;
  border-style: solid;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c0::before {
  border-left-width: 1.5rem;
  border-right-width: 0rem;
  border-left-color: transparent;
  left: 0;
}

<div
  class="c0"
>
  #006
</div>
`;
